### PR TITLE
 Synchronous state machine

### PR DESCRIPTION
**Pull request by andersio**
Depends on #7.

1. Represent the state machine as `Property` instead of `SignalProducer`.
   ```swift
   self.state = Property(initial: State.initial,
                         reduce: State.reduce,
                         feedbacks: feedbacks)
   ```

2. Drop the mandatory scheduler requirement.
    The state machine has now an unbound event queue that flattens recursive events.

3. Tweaked the implementation so that only one copy of the state is stored by the state machine.

4. The examples have been updated with the new API, and some minor visual tweaks.

### CODE REVIEW

**Comment by sergdort** [comment]
Interesting =) Does it pass [this](https://github.com/Babylonpartners/ReactiveFeedback/pull/3/files#diff-2aac158374aba768ff3e5a371d5affcdR52) test tho?

**Comment by andersio** [comment]
@sergdort No. Now that the system is a hot subject, the test constitutes an infinite hot feedback loop. It won't deadlock though, just spinning forever.

An adapted version of the test: https://github.com/Babylonpartners/ReactiveFeedback/pull/10/commits/1eab3b6c6757ae2d482cd5fb0cac537c1a7f988f

**Comment by sergdort** [comment]
I see several issues with this approach: 

1. We are no longer in control of the state machine. e.g we can not say `take(first:)` or `takeUntill`, `replayLazyly` etc. Which means are not able to create composable `System`.
2. According to current implementation state machine is running before even `Property` initialized. This is *very important* IMO as somebody who would want to be interested in all changes of the state will lose it. And IMO it this wrong from `Rx` and `ReactiveSwift` idea as `SignalProducer` is deferred piece of work
3. If somebody would want to start the system again they would need to wire everything again. Instead of calling `start` on `SignalProducer`
4. I also feel like this implementation fits only `MVVM` but what if someone would want to do something similar to `Elm` and treat UI as another `FeedbackLoop` and wire everything in `ViewController`

P.S Also I feel like this approach is less elegant (Please don't get me wrong, I'm not defending my implementation as essentially I just translate it to `ReactiveSwift`)

**Comment by sergdort** [comment]
I would not want to put any user of the library into "frames" - "This is only way you can use it"
I would like a users to decide what best fits their needs and architectural approach, so IMHO exposing `SignalProducer` is the best approach.

**Review by sergdort — CHANGES_REQUESTED** [review]
I see your point and I think `Property` can be useful in some scenarios. How about we add convenience method that would return `Property`

File: ReactiveFeedback/Property+StateMachine.swift [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                        return true
+                    }
+
+                    gate.queue.append(event)
+                    return false
+                }
+
+                guard canProceed else { return }
+
+                state.modify { $0 = reduce($0, event) }
```
Comment (sergdort) [inline ReactiveFeedback/Property+StateMachine.swift:54]: Won't it fire events recursively?

Comment (andersio) [reply ReactiveFeedback/Property+StateMachine.swift:54]: There is a FIFO buffer to flatten recursive events into an iterative loop.

Comment (sergdort) [reply ReactiveFeedback/Property+StateMachine.swift:54]: Yeah, but won't it be an issue for the `ReactiveSwift`?

**Comment by andersio** [comment]
> According to current implementation state machine is running before even Property initialized.

Even if it is created as a SignalProducer initially, as you drop it into `Property(initial:then:)` the side effects start immediately.

> This is very important IMO as somebody who would want to be interested in all changes of the state will lose it. 

Whenever you promote a stream to `Property` which has a stronger contract, one loses all values prior to the current state.

For most purposes — at least for the purpose of GUI — the lost history rarely matters.

Let's say why should I care if a screen was `loading` in the past if it is already `loaded`? Generally, only the present and the future is ever concerned, and `Property` covers both.

> And IMO it this wrong from Rx and ReactiveSwift idea as SignalProducer is deferred piece of work

Deferred until you start it. Promotion to `Property` starts it implicitly.

> If somebody would want to start the system again they would need to wire everything again. Instead of calling start on SignalProducer

The actual question is perhaps _why_ anybody would need to do it with finite state machine in mind. If one needs to restart, it smells like a bad design.

> I also feel like this implementation fits only MVVM but what if someone would want to do something similar to Elm and treat UI as another FeedbackLoop and wire everything in ViewController

> I would not want to put any user of the library into "frames"

> We are no longer in control of the state machine. e.g we can not say take(first:) or takeUntill, replayLazyly etc. Which means are not able to create composable System.

A producer variant can always be introduced as needed.

**Comment by sergdort** [comment]
@andersio  Something wrong is going on with the example 
something-wrong

**Comment by andersio** [comment]
@sergdort The bug in the example was fixed. GIF updated. Guess I misunderstood how `newMovies` is supposed to interact with the VC. 💁‍♂️

**Comment by andersio** [comment]
Adding property convenience initialiser would be raised as a separate PR. I would not proceed on the synchronous part, since the time I spent has exceeded my expectation, and to be fair it is functioning well in the spikes.

**Pull request closed without merging**